src/audio/copier/copier.c

```diff
@@ -780,6 +826,178 @@ static int copier_comp_trigger(struct comp_dev *dev, int cmd)
+	 * copier_copy() and here.
+	 */
+
+	avail_frames = audio_stream_get_avail_frames(&multi_buf_c->stream);
+	container_size = audio_stream_sample_bytes(&multi_buf_c->stream);
+
+	avail_bytes = avail_frames * audio_stream_frame_bytes(&multi_buf_c->stream);
+	buffer_stream_invalidate(multi_buf_c, avail_bytes);
+
+	for (endp_idx = 0; endp_idx < cd->endpoint_num; endp_idx++) {
+		uint32_t bytes_produced;
+
+		endp_buf_c = buffer_acquire(cd->endpoint_buffer[endp_idx]);
+
+		for (endp_channel = 0; endp_channel < endp_buf_c->stream.channels;
+				endp_channel++) {
+			int multi_buf_channel = endp_buf_c->chmap[endp_channel];
+
+			if (container_size == 2) {
+				copy_single_channel_c16(&endp_buf_c->stream, endp_channel,
+							&multi_buf_c->stream, multi_buf_channel,
+							avail_frames);
+			} else {
+				assert(container_size == 4);
+				copy_single_channel_c32(&endp_buf_c->stream, endp_channel,
+							&multi_buf_c->stream, multi_buf_channel,
+							avail_frames);
+			}
+		}
+
+		bytes_produced = avail_frames * audio_stream_frame_bytes(&endp_buf_c->stream);
+		buffer_stream_writeback(endp_buf_c, bytes_produced);
+		comp_update_buffer_produce(endp_buf_c, bytes_produced);
+		buffer_release(endp_buf_c);
+	}
+
+	comp_update_buffer_consume(multi_buf_c, avail_bytes);
+	buffer_release(multi_buf_c);
+}
+
+static void mux_into_multi_endpoint_buffer(struct copier_data *cd)
+{
+	struct comp_buffer __sparse_cache *multi_buf_c, *endp_buf_c;
+	int endp_idx, endp_channel;
+	uint32_t avail_frames = UINT32_MAX, container_size, bytes_produced;
+
+	for (endp_idx = 0; endp_idx < cd->endpoint_num; endp_idx++) {
+		endp_buf_c = buffer_acquire(cd->endpoint_buffer[endp_idx]);
+		avail_frames = MIN(avail_frames,
+				   audio_stream_get_avail_frames(&endp_buf_c->stream));
+		buffer_release(endp_buf_c);
+	}
+
+	/* NOTE: free space in destination buffer is not checked. This is done to have
+	 * behaviour consistent with copier_copy() which does not check for free space either.
+	 * However, quite probably, such behaviour is just a bug and have to be fixed in both
+	 * copier_copy() and here.
+	 */
+
+	multi_buf_c = buffer_acquire(cd->multi_endpoint_buffer);
+
+	container_size = audio_stream_sample_bytes(&multi_buf_c->stream);
+
+	for (endp_idx = 0; endp_idx < cd->endpoint_num; endp_idx++) {
+		uint32_t endp_buf_avail_bytes;
+
+		endp_buf_c = buffer_acquire(cd->endpoint_buffer[endp_idx]);
+
+		endp_buf_avail_bytes = avail_frames * audio_stream_frame_bytes(&endp_buf_c->stream);
+		buffer_stream_invalidate(endp_buf_c, endp_buf_avail_bytes);
+
+		for (endp_channel = 0; endp_channel < endp_buf_c->stream.channels;
+				endp_channel++) {
+			int multi_buf_channel = endp_buf_c->chmap[endp_channel];
+
+			if (container_size == 2) {
+				copy_single_channel_c16(&multi_buf_c->stream, multi_buf_channel,
+							&endp_buf_c->stream, endp_channel,
+							avail_frames);
+			} else {
+				assert(container_size == 4);
```

Comment:
Hmm, audio stream config is coming from outside (which is not trusted), so assert() is doesn't seem like the correct option here.

Reply:
Replaced with switch() and error reporting for unsupported container size.